crates/ruff_python_formatter/src/expression/mod.rs

```diff
@@ -154,6 +169,29 @@ impl Format<PyFormatContext<'_>> for MaybeParenthesizeExpression<'_> {
             parenthesize,
         } = self;
 
+        // Fluent style means we always add parentheses, so we don't need to check for existing
+        // parentheses, comments or `needs_parentheses`
+        let is_fluent_style = is_fluent_style_call_chain(expression, f.context().source());
+        if is_fluent_style {
+            match expression {
+                Expr::Attribute(expr) => {
+                    return parenthesize_if_expands(&group(&expr.format().with_options(true)))
+                        .fmt(f)
+                }
+                Expr::Call(expr) => {
+                    return parenthesize_if_expands(&group(&expr.format().with_options(true)))
+                        .fmt(f)
+                }
+                Expr::Subscript(expr) => {
+                    return parenthesize_if_expands(&group(&expr.format().with_options(true)))
+                        .fmt(f)
+                }
```

Comment:
I don't think i really follow; Call chains can appear about everywhere except on statement level. For the MaybeParenthesize it's special because we need to switch parentheses mode, call chain formatting always has parentheses when breaking.

> Change can_omit_parentheses to return false for call chains.

This is already the case.

Unfortunately I don't understand why exactly the current solution works but the alternatives i tried don't, the interaction of node level, context, parentheses, MaybeParenthesize and FormatExpr or not is still opaque to me.